File: vscode/src/core/StateWatcher.ts
/**
 * IME 状态文件监听器
 * 从 Kotlin core/ime/StateWatcher.kt 移植
 * 监听 Rime Lua 脚本写入的当前应用 session 状态文件
 */

import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import { ImeState, Logger } from './types';

const RIME_STATE_FILE_NAME = 'ime-state-rime-v2.json';

/**
 * 解析 Rime session 状态 JSON；不完整写入直接返回 null。
 */
export interface RimeSessionState {
  state: ImeState;
  sessionToken: string;
  sequence: number;
}

export function parseRimeSessionState(content: string): RimeSessionState | null {
  try {
    const data = JSON.parse(content) as Record<string, unknown>;
    if (
      data.protocol_version !== 2 ||
      data.provider !== 'rime' ||
      typeof data.session_token !== 'string' || !data.session_token ||
      typeof data.sequence !== 'number' ||
      !Number.isSafeInteger(data.sequence) || data.sequence < 1 ||
      typeof data.ascii_mode !== 'boolean' ||
      typeof data.caps_lock !== 'boolean' ||
      typeof data.is_composing !== 'boolean' ||
      typeof data.timestamp !== 'number' ||
      !Number.isSafeInteger(data.timestamp) || data.timestamp < 1
    ) {
      return null;
    }

    return {
      state: {
        isAsciiMode: data.ascii_mode,
        isCapsLock: data.caps_lock,
        isComposing: data.is_composing,
      },
      sessionToken: data.session_token,
      sequence: data.sequence,
    };
  } catch {
    return null;
  }
}

export class RimeSessionTracker {
  private currentSessionToken: string | undefined;
  private lastSequence = 0;
  accept(update: RimeSessionState): boolean {
    if (
      update.sessionToken === this.currentSessionToken &&
      update.sequence <= this.lastSequence
    ) {
      return false;
    }

    this.currentSessionToken = update.sessionToken;
    this.lastSequence = update.sequence;
    return true;
  }
}

export class RimeStateWatcher {
  private watcher: fs.FSWatcher | null = null;
  private pollTimer: ReturnType<typeof setInterval> | null = null;
  private isRunning = false;

  isComposing = false;
  lastAsciiMode = true;
  lastCapsLock = false;

  /** 防止递归：正在强制切换 IME 时跳过状态文件读取 */
  isForcingImeSwitch = false;

  private readonly stateFilePath: string;
  private readonly sessionTracker: RimeSessionTracker;

  constructor(
    private logger: Logger,
    private onStateChanged: (state: ImeState) => void,
  ) {
    const fileName = RIME_STATE_FILE_NAME;
    this.stateFilePath = path.join(
      process.env.TEMP || os.tmpdir(),
      fileName,
    );
    this.sessionTracker = new RimeSessionTracker();
  }

  start(): void {
    if (this.isRunning) {
      this.logger.debug('StateWatcher already running');
      return;
    }

    this.logger.info(`Starting StateWatcher, monitoring: ${this.stateFilePath}`);
    this.isRunning = true;

    // fs.watch 用于即时通知，轮询用于兜底 Windows 上可能丢失的文件事件。
    this.startFileWatch();
    this.startPolling();

    // 初始化：读取当前状态
    this.readAndApplyState();
  }

  stop(): void {
    this.isRunning = false;
    if (this.watcher) {
      this.watcher.close();
      this.watcher = null;
    }
    if (this.pollTimer) {
      clearInterval(this.pollTimer);
      this.pollTimer = null;
    }
    this.logger.info('StateWatcher stopped');
  }

  refresh(): void {
    this.readAndApplyState();
  }

  private startFileWatch(): void {
    const dir = path.dirname(this.stateFilePath);
    const fileName = path.basename(this.stateFilePath);

    try {
      this.watcher = fs.watch(dir, (eventType, changedFile) => {
        if (changedFile === fileName) {
          this.readAndApplyState();
        }
      });
      this.watcher.on('error', (err: Error) => {
        this.logger.warn('fs.watch error, continuing with polling', err);
        this.watcher?.close();
        this.watcher = null;
      });
    } catch (e) {
      this.logger.warn('fs.watch not available, using polling', e as Error);
    }
  }

  /**
   * 轮询回退方案（500ms 间隔）
   */
  private startPolling(): void {
    if (this.pollTimer) return;
    this.pollTimer = setInterval(() => {
      if (!this.isRunning) return;
      this.readAndApplyState();
    }, 500);
  }

  private readAndApplyState(): void {
    if (this.isForcingImeSwitch) {
      this.logger.debug('Skipping state file read during forced IME switch');
      return;
    }

    try {
      const update = this.readStateFile(this.stateFilePath);
      if (!update) return;
      if (!this.sessionTracker.accept(update)) return;
      const state = update.state;

      if (
        state.isAsciiMode !== this.lastAsciiMode ||
        state.isCapsLock !== this.lastCapsLock ||
        state.isComposing !== this.isComposing
      ) {
        this.logger.info(
          `IME state changed: ascii=${state.isAsciiMode}, caps=${state.isCapsLock}, composing=${state.isComposing}`,
        );
        this.lastAsciiMode = state.isAsciiMode;
        this.lastCapsLock = state.isCapsLock;
        this.isComposing = state.isComposing;

        this.onStateChanged(state);
      }
    } catch (e) {
      this.logger.debug('Failed to parse state file (may be in progress)');
    }
  }

  private readStateFile(stateFilePath: string): RimeSessionState | null {
    if (!fs.existsSync(stateFilePath)) return null;
    const content = fs.readFileSync(stateFilePath, 'utf-8').trim();
    if (!content) return null;
    return parseRimeSessionState(content);
  }
}

File: core/src/main/kotlin/com/auto_switch_ime/core/ime/StateWatcher.kt
package com.auto_switch_ime.core.ime

import com.auto_switch_ime.core.ImeState
import com.auto_switch_ime.core.util.Logger
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.longOrNull
import java.io.File
import java.nio.file.FileSystems
import java.nio.file.StandardWatchEventKinds
import java.util.concurrent.TimeUnit

/**
 * IME 状态文件监听器
 * 监听 Rime Lua 脚本写入的状态文件变化
 * 平台无关，通过回调通知状态变化
 */
class RimeStateWatcher(
    private val stateFilePath: String,
    private val logger: Logger,
    private val onStateChanged: (ImeState) -> Unit
) {
    private val sessionTracker = RimeSessionTracker()
    private var watcherThread: Thread? = null
    private var isRunning = false

    @Volatile
    var isComposing: Boolean = false
        private set

    @Volatile
    var lastAsciiMode: Boolean = true
        private set

    @Volatile
    var lastCapsLock: Boolean = false
        private set

    // 防止递归：正在强制切换 IME 时不再处理状态文件变化
    @Volatile
    var isForcingImeSwitch: Boolean = false

    /**
     * 启动监听
     */
    fun start() {
        if (isRunning) {
            logger.debug("StateWatcher already running")
            return
        }

        logger.info("Starting StateWatcher, monitoring: $stateFilePath")
        isRunning = true

        watcherThread = Thread({
            watchLoop()
        }, "AutoSwitchIME-StateWatcher").apply {
            isDaemon = true
            start()
        }

        // 初始化：读取当前状态文件（如果存在）
        readAndApplyState()
    }

    /**
     * 停止监听
     */
    fun stop() {
        isRunning = false
        watcherThread?.interrupt()
        watcherThread = null
        logger.info("StateWatcher stopped")
    }

    private fun watchLoop() {
        val stateFile = File(stateFilePath)

        while (isRunning && !Thread.currentThread().isInterrupted) {
            var shouldRestart = false
            var watchService: java.nio.file.WatchService? = null
            try {
                val parentDir = stateFile.parentFile.toPath()
                val service = FileSystems.getDefault().newWatchService()
                watchService = service
                val fileName = stateFile.name

                // 监听父目录
                parentDir.register(
                    service,
                    StandardWatchEventKinds.ENTRY_MODIFY,
                    StandardWatchEventKinds.ENTRY_CREATE
                )

                logger.debug("Watching directory: $parentDir for $fileName changes")

                while (isRunning && !Thread.currentThread().isInterrupted) {
                    val watchKey = service.poll(500, TimeUnit.MILLISECONDS)
                    if (watchKey == null) {
                        readAndApplyState()
                        continue
                    }

                    for (event in watchKey.pollEvents()) {
                        val context = event.context()?.toString() ?: continue
                        if (context == fileName) {
                            readAndApplyState()
                            break
                        }
                    }

                    // 重置 watch key 以继续监听
                    if (!watchKey.reset()) {
                        logger.warn("Watch key no longer valid, restarting watcher")
                        shouldRestart = true
                        break
                    }
                }
            } catch (e: InterruptedException) {
                logger.debug("StateWatcher interrupted")
                Thread.currentThread().interrupt()
            } catch (e: Exception) {
                logger.warn("Error in StateWatcher: ${e.message}", e)
                shouldRestart = isRunning && !Thread.currentThread().isInterrupted
            } finally {
                try {
                    watchService?.close()
                } catch (_: Exception) {
                    // Ignore close failures during watcher restart/shutdown.
                }
            }

            if (shouldRestart && isRunning && !Thread.currentThread().isInterrupted) {
                try {
                    Thread.sleep(1000)
                } catch (e: InterruptedException) {
                    logger.debug("StateWatcher interrupted")
                    Thread.currentThread().interrupt()
                }
            } else {
                break
            }
        }
    }

    fun refresh() {
        readAndApplyState()
    }

    private fun readAndApplyState() {
        // 防止递归
        if (isForcingImeSwitch) {
            logger.debug("Skipping state file read during forced IME switch")
            return
        }

        try {
            val update = readStateFile(stateFilePath) ?: return
            if (!sessionTracker.accept(update)) return
            val state = update.state

            // 检测状态是否发生变化
            if (state.isAsciiMode != lastAsciiMode || state.isCapsLock != lastCapsLock || state.isComposing != isComposing) {
                logger.info(
                    "IME state changed: ascii=${state.isAsciiMode}, caps=${state.isCapsLock}, composing=${state.isComposing}"
                )
                lastAsciiMode = state.isAsciiMode
                lastCapsLock = state.isCapsLock
                isComposing = state.isComposing

                // 通知回调
                onStateChanged(state)
            }
        } catch (e: Exception) {
            // 忽略解析错误（文件可能正在写入中）
            logger.debug("Failed to parse state file (may be in progress): ${e.message}")
        }
    }

    private fun readStateFile(path: String): RimeSessionState? {
        val stateFile = File(path)
        if (!stateFile.exists()) return null

        val content = stateFile.readText(Charsets.UTF_8).trim()
        if (content.isEmpty()) return null
        return parseRimeSessionStateJson(content)
    }

}

internal data class RimeSessionState(
    val state: ImeState,
    val sessionToken: String,
    val sequence: Long
)

internal class RimeSessionTracker {
    private var currentSessionToken: String? = null
    private var lastSequence: Long = 0

    fun accept(update: RimeSessionState): Boolean {
        if (update.sessionToken == currentSessionToken && update.sequence <= lastSequence) return false

        currentSessionToken = update.sessionToken
        lastSequence = update.sequence
        return true
    }
}

internal fun parseRimeSessionStateJson(json: String): RimeSessionState? {
    return try {
        val root = Json.parseToJsonElement(json).jsonObject
        if (root["protocol_version"]?.jsonPrimitive?.intOrNull != 2) return null
        if (root["provider"]?.jsonPrimitive?.contentOrNull != "rime") return null
        val asciiMode = root["ascii_mode"]?.jsonPrimitive?.booleanOrNull ?: return null
        val capsLock = root["caps_lock"]?.jsonPrimitive?.booleanOrNull ?: return null
        val isComposing = root["is_composing"]?.jsonPrimitive?.booleanOrNull ?: return null
        val sessionToken = root["session_token"]?.jsonPrimitive?.takeIf { it.isString }?.contentOrNull
            ?.takeIf { it.isNotBlank() } ?: return null
        val sequence = root["sequence"]?.jsonPrimitive?.longOrNull?.takeIf { it >= 1 } ?: return null
        root["timestamp"]?.jsonPrimitive?.longOrNull?.takeIf { it >= 1 } ?: return null

        RimeSessionState(
            state = ImeState(asciiMode, capsLock, isComposing),
            sessionToken = sessionToken,
            sequence = sequence
        )
    } catch (e: Exception) {
        null
    }
}
